pkg/util/slice_operations.go

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"sync"
+)
+
+type idxRange struct {
+	from int
+	to   int
+}
+
+var idxRangesPool = sync.Pool{
+	New: func() interface{} {
+		return &[]idxRange{}
+	},
+}
+
+// RemoveSliceIndexes takes a slice of elements of any type and a sorted slice of indexes of elements in the former slice.
+// It removes the elements at the given indexes from the given slice, while minimizing the number of copy operations
+// that are required to do so by grouping consecutive indexes into ranges and removing the ranges in single operations.
+// The given number of indexes must be sorted in ascending order.
+// Indexes which are duplicate or out of the range of the given slice are ignored.
+// The returned values are:
+// - The resulting slice of elements after the elements at the given indexes have been removed.
+// - The number of elements that have been removed.
+// - The number of ranges of consecutive elements that have been removed.
```

Comment:
Why is number of ranges interesting for the caller?

Reply:
it's mainly just for testability and there is barely any additional overhead in returning it.

Reply:
Personally I would drop it from public API, but it's no big deal.

Reply:
how would you drop it from the public API while still making sure that the function is testable? Make it private and then wrap it in a public function which drops this return value?

Reply:
> Make it private and then wrap it in a public function which drops this return value?

Yes.

Reply:
how about this db981d937781fabebcaaeb317b75f14c5d58d5c6